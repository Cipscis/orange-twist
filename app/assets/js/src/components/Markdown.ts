import { h } from 'preact';
import htm from 'htm';

import { marked } from 'marked';
import { useEffect, useRef } from 'preact/hooks';
import { JSXInternal } from 'preact/src/jsx.js';

// Initialise htm with Preact
const html = htm.bind(h);

export interface MarkdownProps extends JSXInternal.DOMAttributes<HTMLDivElement> {
	content: string;
}

export function Markdown(props: MarkdownProps) {
	const {
		content,
		...passthroughProps
	} = props;

	const wrapperRef = useRef<HTMLDivElement>(null);

	useEffect(() => {
		const wrapper = wrapperRef.current;
		if (!wrapper) {
			return;
		}

		const renderedContent = marked
			.parse(content, { mangle: false, headerIds: false })
			// Stupid fucking plugin replaces tabs with spaces
			.replace(/ {4}/g, '\t');
		wrapper.setHTML(renderedContent);
	}, [content]);

	return html`<div
		ref="${wrapperRef}"
		class="content"
		...${passthroughProps}
	></div>`;
}
