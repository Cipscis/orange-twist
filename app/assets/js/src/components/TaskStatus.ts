import { h } from 'preact';
import htm from 'htm';

import { Task } from '../types/Task.js';
import { TaskStatus } from '../types/TaskStatus.js';
import { useCallback, useEffect, useState } from 'preact/hooks';
import { setTaskData } from '../registers/tasks/tasksRegister.js';

// Initialise htm with Preact
const html = htm.bind(h);

export interface TaskStatusComponentProps {
	task: Task;
	dayName?: string;
}

const taskStatusSymbols = {
	[TaskStatus.TODO]: '☐',
	[TaskStatus.IN_PROGRESS]: '▶️',
	[TaskStatus.COMPLETED]: '☑️',
} as const satisfies Record<TaskStatus, string>;

export function TaskStatusComponent(props: TaskStatusComponentProps) {
	const {
		task: {
			id,
			status,
		},
		dayName,
	} = props;

	const statusSymbol = taskStatusSymbols[status];

	const [isInChangeMode, setIsInChangeMode] = useState(false);
	const changeStatus = useCallback((newStatus: TaskStatus) => {
		setTaskData(id, { status: newStatus }, { dayName });
		setIsInChangeMode(false);
	}, []);

	// TODO: Turn the selector part into a custom element, using shadow DOM
	const exitChangeModeOnEscape = useCallback((e: KeyboardEvent) => {
		if (e.key === 'Escape') {
			setIsInChangeMode(false);
		}
	}, []);
	useEffect(() => {
		if (isInChangeMode) {
			document.addEventListener('keydown', exitChangeModeOnEscape);
		}

		return () => {
			document.removeEventListener('keydown', exitChangeModeOnEscape);
		};
	}, [isInChangeMode]);

	return html`
		<span class="task-status">
			<button
				type="button"
				title="${status}"
				onClick="${() => setIsInChangeMode(!isInChangeMode)}"
			>${statusSymbol}</button>

			${
				isInChangeMode &&
				html`
					<ul>
						${Object.values(TaskStatus).map((taskStatus) => html`
							<li key="${taskStatus}">
								<button
									type="button"
									onClick="${() => changeStatus(taskStatus)}"
								>
									${taskStatusSymbols[taskStatus]}
								</button>
							</li>
						`)}
					</ul>
				`
			}
		</span>
	`;
}
