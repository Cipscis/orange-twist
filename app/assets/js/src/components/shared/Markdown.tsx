import { h, type JSX } from 'preact';

import { classNames } from 'utils';

import { marked } from 'marked';
import { useLayoutEffect, useRef } from 'preact/hooks';
import { markedHighlight } from 'marked-highlight';
import hljs from 'highlight.js';

interface MarkdownProps extends h.JSX.HTMLAttributes<HTMLDivElement> {
	/**
	 * The markdown content to be rendered as HTML.
	 */
	content: string;
	/**
	 * If set to `true`, only the first line of the content will be used,
	 * and it won't be wrapped in a `<p>` tag.
	 *
	 * @default false
	 */
	inline?: boolean;
}

marked.use({
	renderer: {
		link(href, title, text) {
			try {
				const url = new URL(href, location.origin);
				const isExternal = url.origin !== location.origin;

				return `<a href="${href}"${
					isExternal
						? ' target="_blank"'
						: ''
				}${
					title
						? ` title="${title}"`
						: ''
				}>${text}</a>`;
			} catch (e) {
				// Could not construct a valid URL
				return `<a href="${href}"${
					title
						? ` title="${title}"`
						: ''
				}>${text}</a>`;
			}
		},
	},
});

marked.use(markedHighlight({
	langPrefix: 'hljs language-',
	highlight(code, lang, info) {
		const language = hljs.getLanguage(lang)?.name ?? 'plaintext';
		return hljs.highlight(code, { language }).value;
	},
}));

export function Markdown(props: MarkdownProps): JSX.Element {
	const {
		content,
		inline,
		...passthroughProps
	} = props;

	const wrapperRef = useRef<HTMLDivElement>(null);

	// Using `useLayoutEffect` prevents jittering caused by `setHTML` running after Preact renders
	useLayoutEffect(() => {
		const wrapper = wrapperRef.current;
		if (!wrapper) {
			// The wrapper should always be set by this point
			/* istanbul ignore next */
			return;
		}

		const contentToRender = inline
			? content.split('\n')[0]
			: content;

		(async () => {
			let renderedContent = marked.parse(contentToRender, {
				breaks: true,
			});

			if (typeof renderedContent !== 'string') {
				renderedContent = await renderedContent;
			}

			renderedContent = renderedContent
				// Stupid fucking plugin replaces tabs with spaces
				.replace(/ {4}/g, '\t')
				// To allow HTML tags to be written as text in task names,
				// I replace `<` with `&lt;`. So, when displaying the rendered
				// content, reverse it. This also means if I were type "&amp;lt;"
				// in a task name it would become "<" but that's fine)
				.replace(/&amp;lt;/g, '&lt;');

			if (inline) {
				renderedContent = renderedContent.replace(/<p>(.+)<\/p>\n?/, '$1');
			}

			if (wrapper.setHTML) {
				wrapper.setHTML(renderedContent);
			} else {
				// `setHTML` is not supported, so falling back to vulnerable method'
				wrapper.innerHTML = renderedContent;
			}
		})();
	}, [content, inline]);

	return <div
		ref={wrapperRef}
		{...passthroughProps}
		class={classNames('content', passthroughProps.class && String(passthroughProps.class))}
	/>;
}
