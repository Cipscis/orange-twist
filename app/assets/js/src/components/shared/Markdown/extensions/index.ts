export { taskLink } from './taskLink';
export { template } from './template';
