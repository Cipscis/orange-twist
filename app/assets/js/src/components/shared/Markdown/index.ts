export { Markdown, type MarkdownApi } from './Markdown';
