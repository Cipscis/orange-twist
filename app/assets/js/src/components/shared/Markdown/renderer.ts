import type { RendererObject } from 'marked';

export const renderer: RendererObject = {
	link(href, title, text) {
		try {
			// Render external links in a new tab
			const url = new URL(href, location.origin);
			const isExternal = url.origin !== location.origin;

			return `<a href="${href}"${
				isExternal
					? ' target="_blank"'
					: ''
			}${
				title
					? ` title="${title}"`
					: ''
			}>${text}</a>`;
		} catch (e) {
			// Could not construct a valid URL
			return `<a href="${href}"${
				title
					? ` title="${title}"`
					: ''
			}>${text}</a>`;
		}
	},
};
