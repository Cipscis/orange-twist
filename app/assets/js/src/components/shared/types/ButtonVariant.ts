import type { EnumTypeOf } from 'utils';

export const ButtonVariant = {
	PRIMARY: 'primary',
	SECONDARY: 'secondary',
} as const;
export type ButtonVariant = EnumTypeOf<typeof ButtonVariant>;
