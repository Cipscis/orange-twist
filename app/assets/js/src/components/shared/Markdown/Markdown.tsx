import { h, type JSX } from 'preact';
import { useLayoutEffect, useRef } from 'preact/hooks';

import { classNames } from 'utils';

import { marked, type Tokens } from 'marked';
import { markedHighlight } from 'marked-highlight';
import hljs from 'highlight.js';
import { getTaskInfo } from 'data';
import { TaskStatusSymbol } from 'types/TaskStatus';

interface MarkdownProps extends h.JSX.HTMLAttributes<HTMLDivElement> {
	/**
	 * The markdown content to be rendered as HTML.
	 */
	content: string;
	/**
	 * If set to `true`, only the first line of the content will be used,
	 * and it won't be wrapped in a `<p>` tag.
	 *
	 * @default false
	 */
	inline?: boolean;
}

marked.use({
	renderer: {
		link(href, title, text) {
			try {
				// Render external links in a new tab
				const url = new URL(href, location.origin);
				const isExternal = url.origin !== location.origin;

				return `<a href="${href}"${
					isExternal
						? ' target="_blank"'
						: ''
				}${
					title
						? ` title="${title}"`
						: ''
				}>${text}</a>`;
			} catch (e) {
				// Could not construct a valid URL
				return `<a href="${href}"${
					title
						? ` title="${title}"`
						: ''
				}>${text}</a>`;
			}
		},
	},
	extensions: [
		{
			name: 'taskLink',
			level: 'inline',
			start(src) {
				// Tell Marked when to stop and check for a match
				return src.match(/\[\[\d/)?.index;
			},
			tokenizer(src, tokens) {
				// Expression for complete token, from its start
				// Match a number in double square braces, and collect any following text
				// e.g. "[[48]]"
				const rule = /^(\[\[(\d+)\]\])(.*?)(?:$|\n)/;
				const match = rule.exec(src);
				if (!match) {
					return undefined;
				}

				const token = {
					type: 'taskLink',
					raw: match[0],
					taskId: match[2],
					remainder: match[3],
					tokens: [],
				};
				// Tell the lexer to parse anything in the remainder
				this.lexer.inline(token.remainder, token.tokens);
				return token;
			},
			renderer(token) {
				const taskLink = (() => {
					const taskId = Number(token.taskId);
					if (isNaN(taskId)) {
						// TODO: Needs styling
						return `Invalid task ID ${token.taskId}`;
					}

					const taskInfo = getTaskInfo(taskId);
					if (!taskInfo) {
						// TODO: Needs styling
						return `No task with ID ${taskId}`;
					}

					// TODO: Needs styling
					const taskLink = `<a href="/task?id=${token.taskId}" class="task-link"><span class="task-link__status">${TaskStatusSymbol[taskInfo.status]}</span> <span class="task-link__name">${taskInfo.name}</span></a>`;
					return taskLink;
				})()

				const remainder = this.parser.parseInline(token.tokens ?? []);
				return `${taskLink}${remainder}`;
			},
		},
	],
});

marked.use(markedHighlight({
	langPrefix: 'hljs language-',
	highlight(code, lang, info) {
		const language = hljs.getLanguage(lang)?.name ?? 'plaintext';
		return hljs.highlight(code, { language }).value;
	},
}));

export function Markdown(props: MarkdownProps): JSX.Element {
	const {
		content,
		inline,
		...passthroughProps
	} = props;

	const wrapperRef = useRef<HTMLDivElement>(null);

	// Using `useLayoutEffect` prevents jittering caused by `setHTML` running after Preact renders
	useLayoutEffect(() => {
		const wrapper = wrapperRef.current;
		if (!wrapper) {
			// The wrapper should always be set by this point
			/* istanbul ignore next */
			return;
		}

		const contentToRender = inline
			? content.split('\n')[0]
			: content;

		(async () => {
			let renderedContent = marked.parse(contentToRender, {
				breaks: true,
			});

			if (typeof renderedContent !== 'string') {
				renderedContent = await renderedContent;
			}

			renderedContent = renderedContent
				// Stupid fucking plugin replaces tabs with spaces
				.replace(/ {4}/g, '\t')
				// To allow HTML tags to be written as text in task names,
				// I replace `<` with `&lt;`. So, when displaying the rendered
				// content, reverse it. This also means if I were type "&amp;lt;"
				// in a task name it would become "<" but that's fine)
				.replace(/&amp;lt;/g, '&lt;');

			if (inline) {
				renderedContent = renderedContent.replace(/<p>(.+)<\/p>\n?/, '$1');
			}

			if (wrapper.setHTML) {
				wrapper.setHTML(renderedContent);
			} else {
				// `setHTML` is not supported, so falling back to vulnerable method'
				wrapper.innerHTML = renderedContent;
			}
		})();
	}, [content, inline]);

	return <div
		ref={wrapperRef}
		{...passthroughProps}
		class={classNames('content', passthroughProps.class && String(passthroughProps.class))}
	/>;
}
