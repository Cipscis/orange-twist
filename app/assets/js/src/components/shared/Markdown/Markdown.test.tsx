import { h } from 'preact';

import {
	afterEach,
	describe,
	expect,
	jest,
	test,
} from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';

import { cleanup, render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';

import { Markdown } from './Markdown';
import { clear, setTaskInfo } from 'data';
import { TaskStatus } from 'types/TaskStatus';

describe('Markdown', () => {
	afterEach(() => {
		cleanup();
		clear();
	});

	test('renders its content, converted to HTML', () => {
		const { getByTestId } = render(
			<Markdown
				content="**Bold** *italic* `code`"
				data-testid="markdown-content"
			/>
		);

		const content = getByTestId('markdown-content');
		expect(content).toBeInTheDocument();
		expect(content.innerHTML.trim()).toBe('<p><strong>Bold</strong> <em>italic</em> <code>code</code></p>');
	});

	test('uses tabs for indentation, not spaces', () => {
		const text = `\`\`\`
	This is some code
\`\`\``;
		const { getByTestId } = render(
			<Markdown
				content={text}
				data-testid="markdown-content"
			/>
		);

		const content = getByTestId('markdown-content');
		expect(content.innerHTML.trim()).toBe('<pre><code>\tThis is some code\n</code></pre>');
	});

	test('converts "&amplt;" into "&lt;"', () => {
		const { getByText } = render(
			<Markdown content="&amp;lt;example>" />
		);

		const content = getByText(/<example>/);
		expect(content).toBeInTheDocument();
	});

	test('renders external links with `target="_blank"`', () => {
		const { getByText } = render(
			<Markdown content={`[Internal link](/test/ "Title")\n[External link](https://example.com "Title")`} />
		);

		const internalLink = getByText('Internal link');
		expect(internalLink).toBeInTheDocument();
		expect(internalLink.getAttribute('href')).toBe('/test/');
		expect(internalLink.getAttribute('target')).toBe(null);
		expect(internalLink.getAttribute('title')).toBe('Title');

		const externalLink = getByText('External link');
		expect(externalLink).toBeInTheDocument();
		expect(externalLink.getAttribute('href')).toBe('https://example.com');
		expect(externalLink.getAttribute('target')).toBe('_blank');
		expect(externalLink.getAttribute('title')).toBe('Title');
	});

	test('allows line breaks without any extra syntax', () => {
		const { getByTestId } = render(
			<Markdown
				content={`First line\nSecond line`}
				data-testid="markdown-content"
			/>
		);

		const content = getByTestId('markdown-content');
		expect(content.innerHTML).toBe('<p>First line<br>Second line</p>\n');
	});

	test('passes through additional props to the internal `<div>`, adding CSS classes to its existing ones', async () => {
		const user = userEvent.setup();
		const spy = jest.fn();

		const { getByTestId, rerender } = render(
			<Markdown
				content=""
				data-testid="markdown-content"
			/>
		);

		let content = getByTestId('markdown-content');
		expect([...content.classList.values()]).toEqual(['content']);

		rerender(
			<Markdown
				content=""
				class="test"
				onClick={spy}
				data-testid="different-test-id"
			/>
		);

		content = getByTestId('different-test-id');
		expect(content.classList.contains('test')).toBe(true);

		expect(spy).not.toHaveBeenCalled();
		await user.click(content);
		expect(spy).toHaveBeenCalledTimes(1);
	});

	test('sanitises content', () => {
		const { getByTestId } = render(
			<Markdown
				content="<script>window.alert('xss')</script>"
				data-testid="markdown-content"
			/>
		);

		const content = getByTestId('markdown-content');
		expect(content.textContent).toBe('window.alert(\'xss\')');
	});

	test('displays the first line of content only, unwrapped, when using the inline prop', () => {
		const content = `This is **the** *first* \`line\`

this is the second line`;

		const { getByTestId } = render(
			<Markdown
				content={content}
				inline
				data-testid="markdown-content"
			/>
		);

		const contentEl = getByTestId('markdown-content');
		expect(contentEl.innerHTML).toBe('This is <strong>the</strong> <em>first</em> <code>line</code>');
	});

	describe('renders task links', () => {
		test('for a valid task ID', () => {
			setTaskInfo(1, {
				name: 'Task name',
				status: TaskStatus.IN_PROGRESS,
			});

			const { getByTestId } = render(
				<Markdown
					content="Here's a task link: [[1]] *what do you think?*"
					data-testid="markdown-content"
				/>
			);

			const content = getByTestId('markdown-content');
			expect(content).toMatchSnapshot();
		});

		test('for an invalid task ID', () => {
			const { getByTestId } = render(
				<Markdown
					content="Here's a task link: [[1]] *what do you think?*"
					data-testid="markdown-content"
				/>
			);

			const content = getByTestId('markdown-content');
			expect(content).toMatchSnapshot();
		});
	});
});
