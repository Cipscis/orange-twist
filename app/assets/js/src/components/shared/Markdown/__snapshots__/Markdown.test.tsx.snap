// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown renders task links for a valid task ID 1`] = `
<div
  class="content"
  data-testid="markdown-content"
>
  <p>
    Here's a task link: 
    <a
      class="task-link"
      href="/task?id=1"
    >
      <span
        class="task-link__status"
      >
        ▶️
      </span>
      <span
        class="task-link__name"
      >
        Task name
      </span>
    </a>
     
    <em>
      what do you think?
    </em>
  </p>
  

</div>
`;

exports[`Markdown renders task links for an invalid task ID 1`] = `
<div
  class="content"
  data-testid="markdown-content"
>
  <p>
    Here's a task link: 
    <span>
      No task with ID $1
    </span>
     
    <em>
      what do you think?
    </em>
  </p>
  

</div>
`;
